chore: Multibackend alerts in auth flow - WPB-17803

dmitrysimkin opened the pull request:
<!--do not remove this marker, its needed to replace info when ticket title is updated -->
<!--jira-description-action-hidden-marker-start-->

<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-17803" title="WPB-17803" target="_blank"><img alt="Task" src="https://wearezeta.atlassian.net/rest/api/2/universal_avatar/view/type/issuetype/avatar/10818?size=medium" />WPB-17803</a>  [iOS] Show alert for unresolved api version in authentication flow
  </td></table>
  <br />
 

<!--jira-description-action-hidden-marker-end-->
<!--do not remove the jira markers to link tickets automatically -->


### Issue

Show mutibacked errors in auth flow. There are 2 alerts

<img width="752" alt="Screenshot 2025-06-06 at 17 21 10" src="https://github.com/user-attachments/assets/d7faf285-797b-4339-81ff-e84e6db3df02" />

### Testing

_Describe how to test._

_Optional: attachments like images, videos, etc._

---

### Checklist

- [ ] Title contains a reference JIRA issue number like `[WPB-XXX]`.
- [ ] Description is filled and free of optional paragraphs.
- [ ] Adds/updates automated tests.

---

### UI accessibility checklist

_If your PR includes UI changes, please utilize this checklist:_
- [ ] Make sure you use the API for UI elements that support large fonts.
- [ ] All colors are taken from WireDesign.ColorTheme or constructed using WireDesign.BaseColorPalette.
- [ ] New UI elements have Accessibility strings for VoiceOver.

dmitrysimkin committed:
Add MultiBackend lib

dmitrysimkin committed:
Merge branch 'develop' into chore/add-multibackend-lib-wpb-17925

dmitrysimkin committed:
Fixed test plan

dmitrysimkin committed:
Add Account switcher initial UI

dmitrysimkin committed:
Added more precised style

dmitrysimkin committed:
Improve style + localzied

dmitrysimkin committed:
Add more preview

dmitrysimkin committed:
Format

dmitrysimkin committed:
Format

dmitrysimkin committed:
Integrate new view

dmitrysimkin committed:
Add actions for options

dmitrysimkin committed:
Map accounts

dmitrysimkin committed:
Add action on account selected

dmitrysimkin committed:
Format

dmitrysimkin committed:
Merge branch 'chore/add-reusable-account-switcher-wpb-17805' into chore/account-switcher-self-profile-intergration-WPB-17807

# Conflicts:
#	WireUI/Sources/WireMultiBackendUI/AccountSwitcherView.swift
#	WireUI/Sources/WireMultiBackendUI/OptionView.swift

dmitrysimkin committed:
Fix build error

dmitrysimkin committed:
Code style

dmitrysimkin committed:
Added initial account switcher in auth module

dmitrysimkin committed:
Make private view

dmitrysimkin committed:
Pass accounts publsiher

dmitrysimkin committed:
Add scheme

dmitrysimkin committed:
Fixed wire auth app building

dmitrysimkin committed:
Setup Needle to pass accounts publisher

dmitrysimkin committed:
Move pushing switch accounts from root

dmitrysimkin committed:
Pass real accounts to account switcher

dmitrysimkin committed:
Move to own file

dmitrysimkin committed:
Add Accessibility strings file

dmitrysimkin committed:
Added scheme

dmitrysimkin committed:
Merge branch 'develop' into chore/add-reusable-account-switcher-wpb-17805

# Conflicts:
#	WireUI/Package.swift
#	WireUI/Sources/WireMultiBackendUI/Resources/Localization/en.lproj/Localizable.strings

dmitrysimkin committed:
Added accessibiltiy strings to config

dmitrysimkin committed:
Merge branch 'develop' into chore/add-reusable-account-switcher-wpb-17805

dmitrysimkin committed:
Merge branch 'chore/add-reusable-account-switcher-wpb-17805' into chore/account-switcher-self-profile-intergration-WPB-17807

# Conflicts:
#	WireUI/Sources/WireMultiBackendUI/OptionView.swift

dmitrysimkin committed:
Merge branch 'develop' into chore/account-switcher-self-profile-intergration-WPB-17807

# Conflicts:
#	WireUI/Sources/WireMultiBackendUI/AccountSwitcherView.swift
#	WireUI/Sources/WireMultiBackendUI/AccountView.swift
#	WireUI/Sources/WireMultiBackendUI/OptionView.swift

dmitrysimkin committed:
Merge branch 'chore/account-switcher-self-profile-intergration-WPB-17807' into chore/account-switcher-modal-WPB-17806

# Conflicts:
#	WireAuthentication/Sources/WireAuthentication/Components/RootComponent.swift
#	WireAuthentication/Sources/WireAuthentication/Needle.generated.swift
#	WireAuthentication/Sources/WireAuthentication/WireAuthenticationAssembly.swift
#	WireUI/Sources/WireMultiBackendUI/AccountView.swift
#	wire-ios/Wire-iOS/Sources/Authentication/AuthenticationInterfaceBuilder.swift

dmitrysimkin committed:
Code style

dmitrysimkin committed:
Format

dmitrysimkin committed:
Add scheme

dmitrysimkin committed:
Show other accounts header + renaming

dmitrysimkin committed:
Fixed buidlig tests

dmitrysimkin committed:
Use dynamic font

dmitrysimkin committed:
Remove not used

dmitrysimkin committed:
Prepare to make PR

dmitrysimkin committed:
Format

dmitrysimkin committed:
Format

dmitrysimkin committed:
Merge branch 'develop' into chore/account-switcher-modal-WPB-17806

# Conflicts:
#	WireUI/Sources/WireMultiBackendUI/AccountSwitcherHostingController.swift
#	WireUI/Sources/WireMultiBackendUI/AccountSwitcherView.swift
#	WireUI/Sources/WireMultiBackendUI/AccountView.swift
#	wire-ios/Wire-iOS/Sources/UserInterface/SelfProfile/SelfProfileViewController.swift

dmitrysimkin committed:
Fixed trailing padding

dmitrysimkin committed:
Merge branch 'develop' into chore/account-switcher-modal-WPB-17806

dmitrysimkin committed:
Fixed tests build error

dmitrysimkin committed:
Format

dmitrysimkin committed:
Pass info about accounts and logged in

dmitrysimkin committed:
Do not use flag providers

caldrian committed:
fix: revert string changes in WireAuthentication - WPB-17409 (#3119)

github-actions[bot] committed:
fix: participants added system message wrong position - WPB-18005 (#3117)

Co-authored-by: Jullian Mercier <31648126+jullianm@users.noreply.github.com>

dmitrysimkin committed:
Revert "Do not use flag providers"

This reverts commit 44392668ebb3c45c4aa7a9a8fc2a51e7962bca87.

dmitrysimkin committed:
Pass flags providers

dmitrysimkin committed:
Code style

dmitrysimkin committed:
Do not pass multibackendEnabled as provider

jullianm committed:
fix: extra dash missed call message - WPB-17974 (#3125)

dmitrysimkin committed:
Pass all accounts to auth flow instead of all other accounts

jullianm committed:
fix: weird transition offline online cherry picks - WPB-17982 (#3132)

jullianm committed:
fix: use wrong user id cherry pick - WPB-18006 (#3133)

dmitrysimkin committed:
Show obsolete client and obsolete server alerts

caldrian committed:
fix: duplicate WireAnalytics symbols - WPB-17530 (#3140)

jullianm committed:
fix: wrong system message user left cherry pick - WPB-17642 (#3130)

github-actions[bot] committed:
chore: Update localization - WPB-17996 (#3143)

dmitrysimkin committed:
Implement log out

dmitrysimkin committed:
PR comments

dmitrysimkin committed:
Remove log out from wire auth where it should not be

dmitrysimkin commented on WireAuthentication/Sources/WireAuthentication/Components/AccountSwitcherComponent.swift:
```diff
@@ -26,7 +26,7 @@ import WireMultiBackendUI
 
     @MainActor var router: any Router { get }
-    var otherAccountsPublisher: ReadOnlyCurrentValueSubject<[AccountUIModel]> { get }
+    var accountsPublisher: ReadOnlyCurrentValueSubject<[AccountUIModel]> { get }
```
As we discussed in Design channel, in auth flow we show all acconuts, not ones excluding current

dmitrysimkin committed:
Fix copies based on feature flag

dmitrysimkin committed:
for client alert message as well

dmitrysimkin committed:
Rename to CurrentValuePublisher

dmitrysimkin committed:
Format

dmitrysimkin committed:
Merge branch 'release/cycle-4.0' into chore/account-switcher-modal-WPB-17806

# Conflicts:
#	WireAuthentication/Package.swift

dmitrysimkin committed:
Merge branch 'chore/account-switcher-modal-WPB-17806' into chore/multibackend-alerts-WPB-17804

samwyndham committed:
fix: read receipts not working - WPB-18016 (#3144)

johnxnguyen commented on WireAuthentication/Sources/WireAuthenticationUI/Views/Root/RootView.swift:
```diff
@@ -80,22 +76,39 @@ package struct RootView: View {
-                message: { Text($0.message) },
-                actions: { alert in
+                title: { alert in
+                    switch alert {
+                    case .obsoleteClient:
+                        Text(
+                            viewModel.multibackendEnabled ? L10n.Localizable.ObsoleteClientMultibackend.Alert
+                                .title : L10n.Localizable.ObsoleteClient.Alert.title
+                        )
+                    default:
+                        Text(alert.title)
+                    }
```
question: since the alerts already have title and message set to them, what do you think about having a new `.multibackendObsoleteClient` and `.multibackendObsoleteServer` alerts and conditionally check the `multibackendEnabled` flag to know which one to present?

johnxnguyen replied:
Why is that? Can't you do the flag check at the point we try to create an alert from an error?

dmitrysimkin replied:
I can do that and it would 2 places where we check `multibackendEnabled` instead of one, it looks to me that it would not bring much value spliting them into separate alerts, wdyt @johnxnguyen ?

johnxnguyen commented on wire-ios/Wire-iOS/Sources/Authentication/Helpers/LogOutHelper.swift:
```diff
@@ -0,0 +1,98 @@
+import WireDataModel
+import WireSyncEngine
+
+class LogOutHelper {
```
nitpick: `final`

dmitrysimkin replied:
Fixed in ebc8fd9ebc2072b33bc007848bdf31bf27c9ee95

dmitrysimkin commented on wire-ios/Wire-iOS/Sources/Authentication/AuthenticationInterfaceBuilder.swift:
```diff
@@ -106,9 +102,9 @@ final class AuthenticationInterfaceBuilder {
-                existsAnotherAccount: numberOfAccounts > 0,
-                otherAccountsPublisher: ReadOnlyCurrentValueSubject(subject: CurrentValueSubject(otherAccounts)),
-                useLegacyRegistrationFlow: !DeveloperFlag.newRegistration.isOn
+                accountsPublisher: CurrentValuePublisher(subject: CurrentValueSubject(accounts)),
```
yes, as talked 1:1 to create a subject in Session Manager, that was my understanding. Something changed since?

johnxnguyen replied:
Probably AccountManager should provide the publisher since it will need to send updates through the subject.

dmitrysimkin committed:
chore: Account switcher modal - WPB-17806 (#3103)

dmitrysimkin committed:
Merge branch 'release/cycle-4.0' into chore/multibackend-alerts-WPB-17804

# Conflicts:
#	WireAuthentication/Sources/WireAuthentication/Components/AccountSwitcherComponent.swift
#	WireAuthentication/Sources/WireAuthentication/Components/RootComponent.swift
#	WireAuthentication/Sources/WireAuthentication/Needle.generated.swift
#	WireAuthentication/Sources/WireAuthentication/WireAuthenticationAssembly.swift
#	WireAuthentication/Sources/WireAuthenticationUI/Resources/Localization/en.lproj/Localizable.strings
#	WireAuthentication/Sources/WireAuthenticationUI/Views/AccountSwitcher/AccountSwitcherModalView.swift
#	WireAuthentication/Sources/WireAuthenticationUI/Views/AccountSwitcher/AccountSwitcherModalViewModel.swift
#	WireAuthentication/Sources/WireAuthenticationUI/Views/Root/RootDestination.swift
#	WireAuthentication/Sources/WireAuthenticationUI/Views/Root/RootView.swift
#	WireFoundation/Sources/WireFoundation/ReadOnlyCurrentValueSubject.swift
#	WirePreviewApps/WireAuthenticationApp/ContentView.swift
#	WireUI/Sources/WireMultiBackendUI/WireMultiBackendUI.swift
#	wire-ios/Wire-iOS/Sources/Authentication/AuthenticationInterfaceBuilder.swift

jullianm committed:
fix: app recovery update events error - WPB-17609 (#3152)

dmitrysimkin committed:
Merge branch 'release/cycle-4.0' into chore/multibackend-alerts-WPB-17804

johnxnguyen reviewed (commented):
Looks good, left some replies to the previous comments

caldrian committed:
chore: update localizations - WPB-17996 (#3175)

jullianm committed:
fix: push notification mute condition issue - WPB-18115 (#3156)

jullianm committed:
fix: unread badges count - WPB-17699 (#3165)

jullianm committed:
fix: missing sync bar - WPB-17717 (#3168)

jullianm reviewed (approved):
Looks good 👍

dmitrysimkin committed:
Merge branch 'release/cycle-4.0' into chore/multibackend-alerts-WPB-17804

jullianm committed:
chore: sync logging cherry pick - WPB-16000 (#3178)

Co-authored-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>

dmitrysimkin committed:
Merge branch 'release/cycle-4.0' into chore/multibackend-alerts-WPB-17804

jullianm committed:
fix: missing messages sync - WPB-18021 (#3173)

github-actions[bot] committed:
fix: detecting urls in messages - WPB-18146 (#3191)

Co-authored-by: Christoph Aldrian <caldrian@users.noreply.github.com>

jullianm committed:
fix: infinite loop - WPB-18213 (#3193)

dmitrysimkin committed:
Merge branch 'release/cycle-4.0' into chore/multibackend-alerts-WPB-17804

caldrian committed:
chore: Multibackend alerts in auth flow - WPB-17803

# Conflicts:
#	WireAuthentication/Sources/WireAuthentication/Components/RootComponent.swift
#	WireAuthentication/Sources/WireAuthentication/WireAuthenticationAssembly.swift
#	wire-ios/Wire-iOS/Sources/Authentication/AuthenticationInterfaceBuilder.swift

caldrian committed:
chore: Multibackend alerts in auth flow - WPB-17803

caldrian committed:
Merge branch 'chore/multibackend-alerts-WPB-17804' into chore/multibackend-alerts-WPB-17804.bak

caldrian committed:
Merge branch 'chore/multibackend-alerts-WPB-17804.bak' into chore/multibackend-alerts-WPB-17804
